### PR TITLE
Better React 18 support

### LINKED ISSUE #3492: An exception is thrown when using Mobx6 and react18 in development mode
**Intended outcome:**

When I use React18's new rendering API `ReactDOM.createRoot().render`, I get a lot of warnings. It should be guaranteed to be consistent with the previous behavior, run normally, and not throw warnings

warning message: react_devtools_backend.js:3973 Warning: Can't perform a React state update on a component that hasn't mounted yet. This indicates that you have a side-effect in your render function that asynchronously later calls tries to update the component. Move this work to useEffect instead.

**Actual outcome:**

I get a lot of warnings, but can't find the information I have in it, like below

**Versions**

mobx@^6.6.1 mobx-react@^7.5.1

Dzdw92s8dG

### PR DESCRIPTION
**Pull request by urugator**
### mobx
It now keeps track of a global state version, which updates with each mutation.

### mobx-react-lite
It now uses `useSyncExternalStore`, which should get rid of tearing (you have to update mobx, otherwise it should behave as previously).
Replaced reaction tracking utils with `UniversalFinalizationRegistry`. It works the same way, but I found it hard to orient myself in the original impl, so I rewrote it completely, hopefully for the better. It's also easier to reuse for class components.
Improved displayName/name handling as discussed in #3438.

### mobx-react (class component)
Reactions of uncommited components are now correctly disposed. (fixes #3492)
Reactions don't notify uncommited components, avoiding the warning. (fixes #3492)
Removed symbol "polyfill" and replaced with actual Symbols.
Removed `this.render` replacement detection + warning. `this.render` is no longer configurable/writable (possibly BC *).
Reaction is no longer exposed as `component[$mobx]`  (possibly BC *)
Component instance is no longer exposed as `component[$mobx]["reactComponent"]` (possibly BC *)
Deprecated `disposeOnUnmount`, it's not compatible with remounting.
Refactored code.
Fixed tests.

(*) BC for non-idiomatic usage or when depending on low level (private?) API.

I will update changeset once the changes settles.

### CODE REVIEW

File: packages/mobx-react/__tests__/__snapshots__/observer.test.tsx.snap [hunk 1]
```diff
@@ -35,23 +37,7 @@ exports[`Redeclaring an existing observer component as an observer should log a
 }
 `;
 
-exports[`SSR works #3448 1`] = `
```
Comment (urugator) [inline packages/mobx-react/__tests__/__snapshots__/observer.test.tsx.snap:38]: Was not supposed to warn. Warning was caused by calling `disableStaticRendering` before `unmount` in the test. Therefore `componentWillUnmount` used logic for non-static rendering.

File: packages/mobx-react/__tests__/inject.test.tsx [hunk 2]
```diff
@@ -102,7 +102,7 @@ describe("inject based context", () => {
         expect(C.displayName).toBe("inject(ComponentC)")
     })
 
-    test.only("shouldn't change original displayName of component that uses forwardRef", () => {
```
Comment (urugator) [inline packages/mobx-react/__tests__/inject.test.tsx:105]: Was left in here for quite some time by mistake.

File: packages/mobx-react/src/observerClass.ts [hunk 3]
```diff
@@ -50,142 +90,156 @@ export function makeClassComponentObserver(
...
-    /**
-     * If props are shallowly modified, react will render anyway,
-     * so atom.reportChanged() should not result in yet another re-render
-     */
-    setHiddenProp(this, skipRenderKey, false)
-    /**
-     * forceUpdate will re-assign this.props. We don't want that to cause a loop,
-     * so detect these changes
-     */
-    setHiddenProp(this, isForcingUpdateKey, false)
```
Comment (urugator) [inline packages/mobx-react/src/observerClass.ts:127]: Merged these together into `isUpdating` as they basically handle the same thing:
One is set by reaction to prevent reportChanged on `props` setter,
second is set by `props` setter to prevent reaction.
So both says "I am already updating, do not cause another update"

File: packages/mobx-react-lite/src/utils/UniversalFinalizationRegistry.ts [hunk 4]
```diff
@@ -0,0 +1,62 @@
+export declare class FinalizationRegistryType<T> {
```
Comment (mweststrate) [inline packages/mobx-react-lite/src/utils/UniversalFinalizationRegistry.ts:1]: Would it be possible to do the FinalizationRegistery refactoring (makes lot of sense) in a separate PR? Then we could first rollout that as a patch change as, as we don't expect changes, and then separately roll out the change for React 18, to reduce the size of the change we have to debug if something goes awry? But if the changes are too intertwined I can live with this as well.

File: packages/mobx-react-lite/src/utils/UniversalFinalizationRegistry.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    private scheduleSweep() {
+        if (this.sweepTimeout === undefined) {
+            this.sweepTimeout = setTimeout(this.sweep, REGISTRY_SWEEP_INTERVAL)
+        }
+    }
+}
+
+export const UniversalFinalizationRegistry = FinalizationRegistry ?? TimerBasedFinalizationRegistry
```
Comment (mweststrate) [inline packages/mobx-react-lite/src/utils/UniversalFinalizationRegistry.ts:62]: Neat setup and implementation!

File: packages/mobx-react/src/disposeOnUnmount.ts [hunk 6]
```diff
@@ -17,12 +20,22 @@ function runDisposersOnWillUnmount() {
...
+/**
+ * @deprecated
+ */
 export function disposeOnUnmount<TF extends Disposer | Array<Disposer>>(
     target: React.Component<any, any>,
     fn: TF
 ): TF
 
+/**
+ * @deprecated
```
Comment (mweststrate) [inline packages/mobx-react/src/disposeOnUnmount.ts:37]: NIT: Let's clarify why it is deprecated, as that shows up in VSCode suggestions

File: packages/mobx/src/core/atom.ts [hunk 7]
```diff
@@ -66,6 +67,9 @@ export class Atom implements IAtom {
     public reportChanged() {
         startBatch()
         propagateChanged(this)
+        // We could update state version only at the end of batch,
+        // but we would still have to switch some global flag here to signal a change.
+        globalState.stateVersion = Symbol()
```
Comment (mweststrate) [inline packages/mobx/src/core/atom.ts:72]: Why not increase a global number? Symbol() is probably a lot more expensive (both mem and cpu), given that this is probably hit very often

File: packages/mobx-react-lite/package.json [hunk 8]
```diff
@@ -36,7 +36,9 @@
         "url": "https://github.com/mobxjs/mobx/issues"
     },
     "homepage": "https://mobx.js.org",
-    "dependencies": {},
+    "dependencies": {
+        "use-sync-external-store": "^1.2.0"
+    },
     "peerDependencies": {
         "mobx": "^6.1.0",
```
Comment (mweststrate) [inline packages/mobx-react-lite/package.json:43]: peer dependency probably has to be bumped to make sure globalState.stateVersion is available?

File: packages/mobx/__tests__/v5/base/observables.js [hunk 9]
```diff
@@ -2360,3 +2360,30 @@ describe("`requiresObservable` takes precedence over global `reactionRequiresObs
         expect(consoleWarnSpy).not.toHaveBeenCalled()
     })
 })
+
+test("state version updates correctly", () => {
+    // This test was designed around the idea of updating version only at the end of batch,
+    // which is NOT an implementation we've settled on, but the test is still valid.
```
Comment (mweststrate) [inline packages/mobx/__tests__/v5/base/observables.js:2366]: Sounds correct to me though :)

**Review by mweststrate — COMMENTED** [review]
Looking good so far! I didn't review everything yet (still have to do the new observer implementations), but submitting this review already so that you can see the comments, as coming days I probably won't be able to review further with all Christmas activities :).

Merry Christmas and happy new year!

File: packages/mobx/src/core/atom.ts (hunk 7 above)
Comment (ericmasiello) [reply packages/mobx/src/core/atom.ts:72]: What makes you say `Symbol()` is more expensive?

**Comment by mweststrate** [comment]
Every Symbol is unique, so at least requires an allocation and GC.

File: packages/mobx/src/core/atom.ts (hunk 7 above)
Comment (urugator) [reply packages/mobx/src/core/atom.ts:72]: > Why not increase a global number?

Becomes non-unique at some point. Dunno how much of an issue it would be in practice. For reference https://github.com/mobxjs/mobx/pull/2982

File: packages/mobx-react-lite/package.json (hunk 8 above)
Comment (urugator) [reply packages/mobx-react-lite/package.json:43]: The current impl should be BC, see
https://github.com/mobxjs/mobx/pull/3590/files#diff-6b93715e19eb9119588dbe2af609a132062e61484d0d3de41b9e32e2d4a4ad5cR13-R17
https://github.com/mobxjs/mobx/pull/3590/files#diff-6b93715e19eb9119588dbe2af609a132062e61484d0d3de41b9e32e2d4a4ad5cR31-R34
https://github.com/mobxjs/mobx/pull/3590/files#diff-6b93715e19eb9119588dbe2af609a132062e61484d0d3de41b9e32e2d4a4ad5cR76-R78

but bumping dep is a way too.

File: packages/mobx/__tests__/v5/base/observables.js (hunk 9 above)
Comment (urugator) [reply packages/mobx/__tests__/v5/base/observables.js:2366]: I am not sure what you mean. Just to provide some background. I've been looking for a common place where I could update state version. `endBatch` (when nesting level gets back to zero) seemed like a good candidate, because it's actually called even for individual mutations, but I was worried about some edge cases, like autorun invoking itself. So I wrote this test and gave it a go. However I quickly realized `endBatch` is also often called without any actual mutation, therefore I threw the idea away and moved the version update to `reportChanged`, but I kept the test as is. I've left the comment to clarify why the test looks this way.

File: packages/mobx-react/src/disposeOnUnmount.ts (hunk 6 above)
Comment (urugator) [reply packages/mobx-react/src/disposeOnUnmount.ts]: Done, but message probably isn't entirely accurate. I am no sure which react version started to perform re-mounting. Maybe it's already done by 17 only in StrictMode, but I don't remember.

File: packages/mobx-react-lite/src/utils/UniversalFinalizationRegistry.ts (hunk 4 above)
Comment (urugator) [reply packages/mobx-react-lite/src/utils/UniversalFinalizationRegistry.ts:1]: PR is here https://github.com/mobxjs/mobx/pull/3598. However I've also changed `useObserver` impl a bit.

File: packages/mobx-react-lite/package.json [hunk 10]
```diff
@@ -36,7 +36,9 @@
         "url": "https://github.com/mobxjs/mobx/issues"
     },
     "homepage": "https://mobx.js.org",
-    "dependencies": {},
+    "dependencies": {
+        "use-sync-external-store": "^1.2.0"
```
Comment (ericmasiello) [inline packages/mobx-react-lite/package.json:40]: Isn't this built into React 18? I'm not sure if there's a way to handle this, but if the user installs React 18, ideally we wouldn't require they also install this, no?

Comment (urugator) [reply packages/mobx-react-lite/package.json:40]: We would have to maintaing seperate mobx-react versions (for >= 18 and < 18), which isn't worth it.
Note the shim prefers React's impl if available. Dunno if it's tree shakable, but it's not very big.

Comment (ericmasiello) [reply packages/mobx-react-lite/package.json:40]: I was hoping something in https://docs.npmjs.com/cli/v7/configuring-npm/package-json#peerdependenciesmeta would solve this problem but I'm not seeing anything unfortunately

File: packages/mobx-react/__tests__/finalizationRegistry.tsx [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+
+afterEach(cleanup)
+
+function sleep(time: number) {
+    return new Promise<void>(res => {
+        setTimeout(res, time)
+    })
+}
+
+// TODO dunno why it's not available
```
Comment (tonyraoul) [inline packages/mobx-react/__tests__/finalizationRegistry.tsx:17]: project was missing a configuration file, please take a look https://github.com/mobxjs/mobx/pull/3620
or simply add the file manually.

File: packages/mobx-react/src/observerClass.ts [hunk 12]
```diff
@@ -1,44 +1,84 @@
-import { PureComponent, Component } from "react"
+import { PureComponent, Component, ComponentClass, ClassAttributes } from "react"
 import {
     createAtom,
     _allowStateChanges,
     Reaction,
-    $mobx,
+    //$mobx, // TODO
```
Comment (mweststrate) [inline packages/mobx-react/src/observerClass.ts:6]: // TODO reminder

**Review by mweststrate — COMMENTED** [review]
Looking great @urugator! Some small comments around publishing, but this looks really solid. Thanks for the extensive comments in the code!

**Review by mweststrate — COMMENTED** [review]
Looking great @urugator! Some small comments around publishing, but this looks really solid. Thanks for the extensive comments in the code!

File: packages/mobx-react-lite/package.json (hunk 8 above)
Comment (urugator) [reply packages/mobx-react-lite/package.json:43]: If this is gonna be a major should it still be focused on React 18 support with otherwise minimal changes, or should this be an opportunity to introduce larger changes? (removing deprecated APIs, hooks, inject, options, cleanups, etc ...)

I assume the former.

**Review by mweststrate — APPROVED** [review]

File: packages/mobx-react-lite/package.json (hunk 10 above)
[the remark this answers is not in the document]
Comment (urugator) [reply packages/mobx-react-lite/package.json:40]: Hm, I think you're correct, did not occured to me.

File: packages/mobx/src/core/atom.ts (hunk 7 above)
Comment (urugator) [reply packages/mobx/src/core/atom.ts]: Changed to number for now. The only thing I am concerned is that if this should ever become a problem, nobody would probably ever notice anyway... not during development at least.

**Comment by ericmasiello** [comment]
I’m curious, how would I observe the “better React 18” support? Can you provide some reproduction steps? For example, I’d love to see something on the order of:
1. Follow these steps with the current version of MobX and React 18 and you can clearly observe *these* negative outputs
2. Repeat the above steps and now you can observe in this branch, those problems are solved.

**Comment by urugator** [comment]
@ericmasiello Besides mentioned fixed issues for which we do have tests, you should see improvements if you run it against:
https://github.com/dai-shi/will-this-react-global-state-work-in-concurrent-rendering
see this comment https://github.com/mobxjs/mobx/issues/2526#issuecomment-1248172620
Personally did not try and I don't have any setup for this, so at this point there is no easily reproducible evidence and you're welcome to change that :)

File: packages/mobx-react/package.json [hunk 13]
```diff
@@ -36,10 +36,10 @@
     },
     "homepage": "https://mobx.js.org",
     "dependencies": {
-        "mobx-react-lite": "^3.4.0"
+        "mobx-react-lite": "^3.4.3"
     },
     "peerDependencies": {
-        "mobx": "^6.1.0",
+        "mobx": "^6.9.0",
```
Comment (urugator) [inline packages/mobx-react/package.json:42]: A guess - I don't know the actual version it's gonna be released with. Dunno what's the correct way to handle it.

File: packages/mobx-react/package.json [hunk 14]
```diff
@@ -51,8 +51,9 @@
         }
     },
     "devDependencies": {
-        "mobx": "^6.7.0",
-        "mobx-react-lite": "^3.4.0"
+        "mobx": "^6.8.0",
+        "mobx-react-lite": "^3.4.3",
```
Comment (urugator) [inline packages/mobx-react/package.json:55]: Should become 6.9?.0 and 4.0.0 at the time of release. Again not sure how to handle that.

File: packages/mobx-react/package.json [hunk 15]
```diff
@@ -36,10 +36,10 @@
     },
     "homepage": "https://mobx.js.org",
     "dependencies": {
-        "mobx-react-lite": "^3.4.0"
+        "mobx-react-lite": "^3.4.3"
```
Comment (urugator) [inline packages/mobx-react/package.json:39]: Should become ^4.0.0 at the time of release. Again not sure how to handle that.

**Comment by urugator** [comment]
@mweststrate Please take a look at https://github.com/mobxjs/mobx/pull/3649 and let me know if I should merge it to master first and incorporate the change to this PR.

**Comment by mweststrate** [comment]
Sorry for the delay! I'll dive into it this weekend

**Review by mweststrate — APPROVED** [review]
Looking good! Imho it's okay to merge #3649 after this one.

File: packages/mobx-react-lite/package.json (hunk 8 above)
Comment (mweststrate) [reply packages/mobx-react-lite/package.json]: I think we can do the former; decorators are also around the corner, an I expect after that is a better moment to do a bigger clean up (e.g. legacy decorator support etc)

**Review by mweststrate — APPROVED** [review]
AWESOME STUFF!

**Comment by ericmasiello** [comment]
I pulled this repo locally and configured it to use 

I see *mostly* progress with 1 regression.  In the output below, "mobx" is `mobx` 6.8.0 and `mobx-react-lite` 3.4.0. "mobx-beta" uses the latest from this branch.

```
  mobx
    With useTransition
      Level 1
        ✓ No tearing finally on update (4571 ms)
        ✓ No tearing finally on mount (5472 ms)
      Level 2
        ✕ No tearing temporarily on update (8641 ms)
        ✕ No tearing temporarily on mount (5460 ms)
      Level 3
        ✓ Can interrupt render (time slicing) (3586 ms)
        ✕ Can branch state (wip state) (2952 ms)
    With useDeferredValue
      Level 1
        ✓ No tearing finally on update (9622 ms)
        ✓ No tearing finally on mount (6486 ms)
      Level 2
        ✓ No tearing temporarily on update (14596 ms)
        ✕ No tearing temporarily on mount (5468 ms)
  mobx-beta
    With useTransition
      Level 1
        ✓ No tearing finally on update (7975 ms)
        ✓ No tearing finally on mount (4555 ms)
      Level 2
        ✓ No tearing temporarily on update (12924 ms)
        ✓ No tearing temporarily on mount (4514 ms)
      Level 3
        ✕ Can interrupt render (time slicing) (7911 ms)
        ✕ Can branch state (wip state) (6685 ms)
    With useDeferredValue
      Level 1
        ✓ No tearing finally on update (9647 ms)
        ✓ No tearing finally on mount (4535 ms)
      Level 2
        ✓ No tearing temporarily on update (14596 ms)
        ✓ No tearing temporarily on mount (4494 ms)
```

**Comment by ericmasiello** [comment]
> @ericmasiello Besides mentioned fixed issues for which we do have tests, you should see improvements if you run it against: https://github.com/dai-shi/will-this-react-global-state-work-in-concurrent-rendering see this comment [#2526 (comment)](https://github.com/mobxjs/mobx/issues/2526#issuecomment-1248172620) Personally I did not try and I don't have any setup for this, so at this point there is no easily reproducible evidence and you're welcome to change that :)

FYI, I have an open PR to add the latest release of mobx/mobx-react-lite to this repo https://github.com/dai-shi/will-this-react-global-state-work-in-concurrent-rendering/pull/75

**Comment by zhantx** [comment]
Hi @urugator 
can you please elaborate why disposeOnUnmount is deprecated? 
I tested it in react 18, it seems the function version is working fine if it's in WillMount (so the reaction is recreated when remount)
Is it only the decorator version problematic in react 18? (I tested it too, it seems can't dispose the reaction properly)

**Comment by upsuper** [comment]
Just want to mention that `disposeOnUnmount` is very useful for us, and it's widely used in our codebase. I'd also note that this utility doesn't encourage, nor does lack of it discourage the pattern. There exists lots of old code that simply collects disposers in `componentDidMount` and invokes them in `componentWillUnmount`. `disposeOnUnmount` helps simplify the code and ensure that they are disposed properly.

**Comment by upsuper** [comment]
I think it's fine that we maintain our own version of this utility in our codebase. We use it both for reaction/autorun and for other stuff following a similar pattern.

Mainly we want to understand the reason that it's deprecated, so that we know whether our use of it has any fundamental issue, in which case our own version of this utility would suffer the same problem as well.

It's good to know that it isn't that the utility really has anything incompatible with React.

**Comment by urugator** [comment]
A potential source of some trouble is the need to patch `componenWillUnmount`, which we already patch by https://github.com/mobxjs/mobx/blob/main/packages/mobx-react/src/utils/utils.ts#L125
The complexity involved is beyond my current knowledge.
